internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			return diag.Errorf("feature Environments is not available")
+		}
+		return diag.FromErr(err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	if err == nil {
+		fmt.Printf("%T \n %v", environmentIDInt, environmentIDInt)
+	}
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return diag.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	name := d.Get("name").(string)
+	options := gitlab.EditEnvironmentOptions{
+		Name: &name,
+	}
+	if d.HasChange("external_url") {
+		options.ExternalURL = gitlab.String(d.Get("external_url").(string))
+	}
+
+	log.Printf("[DEBUG] Project %s update gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	if _, _, err := client.Environments.EditEnvironment(project, environmentIDInt, &options); err != nil {
+		return diag.Errorf("error editing gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return diag.Errorf("error converting environment ID to int: %v", err)
+	}
```

Comment:
Might be worth to refactor out into it's own function already returning the proper types, ...

Reply:
I believe I have refactored this now correctly from 3 copy pasted places to 1 function in this file